src/groups/mqb/mqbblp/mqbblp_relayqueueengine.h

```diff
@@ -127,16 +127,368 @@ struct RelayQueueEngine_AppState : QueueEngineUtil_AppState {
+        /// Element.
+        mqbi::AppMessage* appState();
+
+        /// Return a non-modifiable reference to the App state associated with
+        /// this Element.
+        const mqbi::AppMessage* appView() const;
+
+        /// Return the GUID associated with this Element.
+        Elements& guid() const;
+        App&      app() const;
+
+        void eraseGuid(Stream& stream);
+        void eraseApp(Apps& apps);
+
+        /// Return true if this Element is associated with the specified
+        /// 'iterator' position in the PushStream.
+        bool isInStream(const PushStream::Stream::iterator& iterator) const;
+
+        /// Return reference to the next Element associated with the same GUID
+        /// or '0' if this is the last Element.
+        Element* next() const;
+
+        /// Return reference to the next Element associated with the same App
+        /// or '0' if this is the last Element.
+        Element* nextInApp() const;
+    };
+
+    Stream d_stream;
+
+    Apps d_apps;
+
+    bdlma::ConcurrentPool* d_pushElementsPool_p;
+
+    PushStream(bdlma::ConcurrentPool* pushElementsPool);
+
+    iterator add(const bmqt::MessageGUID& guid);
+
+    /// Add the specified 'element' to both GUID and App corresponding to the
+    /// 'element' (and specified when constructing the 'element').
+    void add(Element* element);
+
+    /// Remove the specified 'element' from both GUID and App corresponding to
+    /// the 'element' (and specified when constructing the 'element').
+    /// Erase the corresponding App from the PushStream if there is no
+    /// remaining Elements in the App.
+    /// Return the number of remaining Elements in the corresponding GUID.
+    unsigned int remove(Element* element);
+
+    /// Remove all PushStream Elements corresponding to the specified
+    /// 'upstreamSubQueueId'.  Erase each corresponding GUIDs from the
+    /// PushStream with no remaining Elements.
+    /// Erase the corresponding App.
+    unsigned int removeApp(unsigned int upstreamSubQueueId);
+
+    /// Remove all PushStream Elements corresponding to the specified
+    /// 'itApp'.    Erase each corresponding GUIDs from the  PushStream with no
+    /// remaining Elements.
+    /// Erase the corresponding App.
+    unsigned int removeApp(Apps::iterator itApp);
+
+    /// Remove all Elements, Apps, and GUIDs.
+    unsigned int removeAll();
+
+    /// Create new Element associated with the specified 'info',
+    // 'upstreamSubQueueId', and 'iterator'.
+    Element* create(const bmqp::SubQueueInfo& info,
+                    const iterator&           iterator,
+                    const Apps::iterator&     iteratorApp);
+
+    /// Destroy the specified 'element'
+    void destroy(Element* element, bool doKeepGuid);
+};
+
+// ==========================================
+// class RelayQueueEngine_PushStorageIterator
+// ==========================================
+
+class RelayQueueEngine_PushStorageIterator : public mqbi::StorageIterator {
```

Comment:
As mentioned above, should this really be a `StorageIterator`?  Or rather, it's not a `StorageIterator`; it has different semantics.  Is it one for convenience or is there a deeper need for it to be one?